### PR TITLE
Lab2 : Synchronization

### PR DESCRIPTION
**Pull request by UsusCimex**

### CODE REVIEW

File: synchronization/syncList.c [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            pthread_mutex_unlock(&current->next->sync);
+            pthread_mutex_unlock(&current->sync);
+            current = current->next;
+        }
+    }
+    return NULL;
+}
+
+// Функция потока для подсчета пар строк, идущих по убыванию длины
+void* countDecreasingLengthPairs(void* arg) {
```
Comment (k-stolyarov) [inline synchronization/syncList.c:75]: You have race conditions here

File: synchronization/syncList.c [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    // Обмен указателями next
+    Node *temp = node1->next;
+    node1->next = node2->next;
+    node2->next = temp;
+}
+
+// Функция потока для перестановки узлов списка
+void* swapNodesInList(void* arg) {
```
Comment (k-stolyarov) [inline synchronization/syncList.c:123]: and here

File: synchronization/syncList.c [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    Storage storage;
+    storage.first = NULL;
+
+    generateList(&storage, 100);
+
+    pthread_t inc_thread, dec_thread, eq_thread, swap_thread, monitor_thread;
+    pthread_create(&inc_thread, NULL, countIncreasingLengthPairs, (void* )&storage);
+    pthread_create(&dec_thread, NULL, countDecreasingLengthPairs, (void* )&storage);
+    pthread_create(&eq_thread, NULL, countEqualLengthPairs, (void* )&storage);
+    pthread_create(&swap_thread, NULL, swapNodesInList, (void* )&storage);
```
Comment (k-stolyarov) [inline synchronization/syncList.c]: please create 3 swaping threads to have badaboom !

File: synchronization/syncList.c [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            pthread_mutex_unlock(&current->next->sync);
+            pthread_mutex_unlock(&current->sync);
+            current = current->next;
+        }
+    }
+    return NULL;
+}
+
+// Функция потока для подсчета пар строк одинаковой длины
+void* countEqualLengthPairs(void* arg) {
```
Comment (k-stolyarov) [inline synchronization/syncList.c:99]: and here

File: synchronization/syncList.c [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        if (temp->next == node) {
+            temp->next = node->next;
+        }
+    }
+    pthread_mutex_destroy(&node->sync);
+    free(node);
+}
+
+// Функция потока для подсчета пар строк, идущих по возрастанию длины
+void* countIncreasingLengthPairs(void* arg) {
```
Comment (k-stolyarov) [inline synchronization/syncList.c:51]: races detected

**Review by k-stolyarov — CHANGES_REQUESTED** [review]

File: synchronization/syncList.c (hunk 1 above)
Comment (UsusCimex) [reply synchronization/syncList.c:75]: fixed all race conditions

File: synchronization/syncList.c (hunk 3 above)
Comment (UsusCimex) [reply synchronization/syncList.c]: I created and encountered a deadlock problem when threads wait for mutexes that are already captured

Comment (UsusCimex) [reply synchronization/syncList.c]: I realized that there shouldn’t be a "badaboom" here, so I changed the implementation and removed the global mutex

**Comment by k-stolyarov** [comment]
please fix history to avoid redundant changes in code review.

File: synchronization/syncList.c [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    Storage* storage = (Storage*)arg;
+    while (1) {
+        Node* current = storage->first;
```
Comment (k-stolyarov) [inline synchronization/syncList.c:56]: races detected

**Review by k-stolyarov — CHANGES_REQUESTED** [review]

File: synchronization/syncList.c (hunk 6 above)
Comment (UsusCimex) [reply synchronization/syncList.c:56]: And now?

File: synchronization/syncList.c [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    }
+    pthread_mutex_destroy(&node->sync);
+    free(node);
+}
+
+// Функция потока для подсчета пар строк, идущих по возрастанию длины
+void* countIncreasingLengthPairs(void* arg) {
+    Storage* storage = (Storage*)arg;
+    while (1) {
+        pthread_mutex_lock(&storage->first->sync);
```
Comment (k-stolyarov) [inline synchronization/syncList.c:55]: Race preserved. When you get mutex to lock  (storage->first->sync and before actual lock your swapping thread might switch first element  and you'll lock mutex not for the first element.

**Review by k-stolyarov — CHANGES_REQUESTED** [review]

File: synchronization/syncList.c (hunk 7 above)
Comment (UsusCimex) [reply synchronization/syncList.c:55]: fixed

**Review by k-stolyarov — APPROVED** [review]
Accepted

**Pull request merged**